src/Functional.CQS.AOP.IoC.SimpleInjector/TypeExtensions.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Functional.CQS.AOP.IoC.SimpleInjector.Models;
+
+// Ambiguous reference in cref attribute (for Option.Some)
+#pragma warning disable CS0419
+#pragma warning disable 1574
+
+namespace Functional.CQS.AOP.IoC.SimpleInjector
+{
+	/// <summary>
+	/// Extension methods for <see cref="Type"/>.
+	/// </summary>
+    public static class TypeExtensions
+    {
+		/// <summary>
+		/// Analyzes a type and returns a closed generic interface (i.e. typeof(IFoo&lt;int&gt;)) if that type implements the open generic interface (i.e. typeof(IFoo&lt;&gt;)).
+		/// </summary>
+		/// <param name="type">The type to analyze.</param>
+		/// <param name="openGenericInterfaceType">The open generic interface type (i.e. typeof(IFoo&lt;&gt;), typeof(IBar&lt;,&gt;)).</param>
+		/// <returns></returns>
+		public static Option<Type> GetClosedGenericInterfaceTypeFromOpenGenericInterfaceType(this Type type, Type openGenericInterfaceType)
+		{
+			return type.GetClosedGenericInterfaceTypeFromOpenGenericInterfaceTypes(new[] { openGenericInterfaceType });
+		}
+
+		/// <summary>
+		/// Analyzes a type and returns a closed generic interface (i.e. typeof(IFoo&lt;int&gt;)) if that type implements any open generic interface specified (i.e. typeof(IFoo&lt;&gt;)).
+		/// </summary>
+		/// <param name="type">The type to analyze.</param>
+		/// <param name="openGenericInterfaceTypes">The open generic interface types (i.e. typeof(IFoo&lt;&gt;), typeof(IBar&lt;,&gt;)).</param>
+		/// <returns></returns>
+		public static Option<Type> GetClosedGenericInterfaceTypeFromOpenGenericInterfaceTypes(this Type type, IEnumerable<Type> openGenericInterfaceTypes)
+		{
+			return Option.FromNullable(type.IsInterface
+				? type.IsConstructedGenericType && openGenericInterfaceTypes.Contains(type.GetGenericTypeDefinition()) ? type : null
+				: type.GetInterfaces().Where(x => x.IsConstructedGenericType).FirstOrDefault(x => openGenericInterfaceTypes.Contains(x.GetGenericTypeDefinition())));
```

Comment:
`.FirstOrDefault(x => y.Contains(x))` suggests this method should build a HashSet from  `openGenericInterfaceTypes`, to prevent an O(n*m) loop.